# The main entry point of your workflow.
# After configuring, running snakemake -n in a clone of this repository should successfully execute a dry-run of the workflow.

report: "report/workflow.rst"

include: "rules/common.smk"

OUTDIR = config["outdir"]
FASTQDIR = samples["fastqs"]
PREFIX = samples["prefix"]

def get_resource(rule,resource):
    try:
        return config["resources"][rule][resource]
    except KeyError:
        return config["resources"]["default"][resource]

def qc_input(wc): 
    path = f"/{FASTQDIR}/{{sample}}/{PREFIX}_S1_L001_R2_001.fastq.gz"
    return path


# Obtain fastq files for fastqc analysis
def get_r1(wildcards):
    return glob.glob(FASTQDIR + wildcards.sample + '_R1.fastq.gz')

def get_r2(wildcards):
    return glob.glob(FASTQDIR + wildcards.sample + '_R2.fastq.gz') 

rule all:
    input:
        # The first rule should define the default target files
        # Subsequent target rules can be specified below. They should start with all_*.
        expand(["{OUTDIR}/cellranger_count/{sample}/raw_feature_bc_matrix.h5",
                "{OUTDIR}/cellranger_count/{sample}/filtered_feature_bc_matrix.h5",
                "{OUTDIR}/cellranger_count/{sample}/possorted_genome_bam.bam",
                "{OUTDIR}/qc/{sample}/{prefix}.html",
                "{OUTDIR}/qc/{sample}/{prefix}_fastqc.zip"], sample=samples['sample'], OUTDIR=OUTDIR, prefix=PREFIX)


include: "rules/cellranger.smk"
include: "rules/qc.smk"
include: "rules/other.smk"
